compositor_common/src/scene/transition.rs

```diff
@@ -0,0 +1,95 @@
+
+use crate::util::InterpolationState;
+
+use super::builtin_transformations::{BuiltinSpec, FixedPositionLayoutSpec};
+
+#[serde_as]
+#[derive(Debug, Serialize, Deserialize, Clone)]
+pub struct TransitionSpec {
+    pub start: BuiltinSpec,
+    pub end: BuiltinSpec,
+    #[serde_as(as = "DurationMilliSeconds<f64>")]
+    #[serde(rename = "transition_duration_ms")]
+    pub transition_duration: Duration,
+    #[serde(flatten)]
+    pub interpolation: Interpolation,
+}
+
+#[derive(Debug, Serialize, Deserialize, Clone)]
+#[serde(tag = "interpolation", rename_all = "snake_case")]
+pub enum Interpolation {
+    Linear,
+    Spring,
+}
+
+impl Interpolation {
+    pub fn interpolate(&self, state: InterpolationState) -> InterpolationState {
+        match self {
+            Interpolation::Linear => state,
+            Interpolation::Spring => {
+                // TODO: placeholder implementation, this needs to be implemented better
+                if state.0 < 0.2 {
+                    InterpolationState(
+                        f64::powf(state.0 * 5.0, 0.3)
+                            + f64::exp(-state.0 * 14.0) * (f64::sin(10.0 * PI * state.0)),
+                    )
+                } else {
+                    InterpolationState(
+                        1.0 + f64::exp(-state.0 * 14.0) * (f64::sin(10.0 * PI * state.0)),
+                    )
+                }
+            }
+        }
+    }
+}
+
+#[derive(Debug)]
+pub enum TransitionType {
+    FixedPositionLayout(FixedPositionLayoutSpec, FixedPositionLayoutSpec),
+}
+
+impl TransitionType {
+    fn validate(&self) -> Result<(), TransitionValidationError> {
+        match self {
+            TransitionType::FixedPositionLayout(start, end) => {
+                FixedPositionLayoutSpec::validate_transition(start, end)
+            }
+        }
+    }
+}
+
+impl TryFrom<(&BuiltinSpec, &BuiltinSpec)> for TransitionType {
+    type Error = TransitionValidationError;
+
+    fn try_from(value: (&BuiltinSpec, &BuiltinSpec)) -> Result<Self, Self::Error> {
+        let transition = match value {
+            (BuiltinSpec::FixedPositionLayout(s1), BuiltinSpec::FixedPositionLayout(s2)) => {
+                Self::FixedPositionLayout(s1.clone(), s2.clone())
+            }
+            (start, end) => {
+                return Err(TransitionValidationError::IncompatibleStartAndEnd(
+                    start.transformation_name(),
+                    end.transformation_name(),
+                ))
+            }
+        };
+        transition.validate()?;
+        Ok(transition)
+    }
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum TransitionValidationError {
+    #[error("Transition from \"{0}\" to \"{1}\" is not supported.")]
+    IncompatibleStartAndEnd(&'static str, &'static str),
+
+    #[error("Field \"{0}\" in transformation \"{1}\" can not be interpolated.")] // TODO: See docs
+    UnsupportedFieldInterpolation(&'static str, &'static str),
```

Comment:
tuple was intentional to make it short, this error will be constructed over and over again. The same argument applies to any function that takes 2 arguments of the same type. I think for 2 args it's fine, but I will think about an approach that is compact on a call site but avoids this issue

> BTW once we add support for transitions in shader transformations, we won't be able to use &'static here anymore

I don't see why we couldn't but if it's the case then we can change it.

Reply:
I assume that in the case of functions, everyone uses an editor / IDE that displays parameter names on the call side. In case of errors, the editor / IDE has no idea what this should mean, because of the name it's defined.

Hints :)
<img width="578" alt="Screenshot 2023-09-25 at 10 56 43" src="https://github.com/membraneframework/video_compositor/assets/104033489/8f94db7b-7e6d-42cb-99b2-f100e7592cec">

No hints :(
<img width="515" alt="Screenshot 2023-09-25 at 11 01 44" src="https://github.com/membraneframework/video_compositor/assets/104033489/bb5223bd-c8a6-444a-a69d-1c38da6b325f">